Add ORTModelXXX for audio

mht-sharma opened the pull request:
# What does this PR do?

Add `ORTModel` for audio models inference

The PR adds:

- `ORTModelForAudioClassification`: ORT inference for [AutoModelForAudioClassification](https://huggingface.co/docs/transformers/model_doc/auto#transformers.AutoModelForAudioClassification) model
- `ORTModelForCTC`: ORT inference for [AutoModelForCTC](https://huggingface.co/docs/transformers/model_doc/auto#transformers.AutoModelForCTC) model
- `ORTModelForAudioXVector`: ORT inference for [AutoModelForAudioXVector](https://huggingface.co/docs/transformers/model_doc/auto#transformers.AutoModelForAudioXVector) model
- `ORTModelForAudioFrameClassification`: ORT inference for [AutoModelForAudioFrameClassification](https://huggingface.co/docs/transformers/model_doc/auto#transformers.AutoModelForAudioFrameClassification) model

Updates
- `ORTModelForSpeechSeq2Seq`: ORT inference for [AutoModelForSpeechSeq2Seq](https://huggingface.co/docs/transformers/model_doc/auto#transformers.AutoModelForSpeechSeq2Seq) model


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/optimum/pr_774). All of your documentation changes will be reflected on that endpoint.

michaelbenayoun commented on optimum/onnxruntime/modeling_ort.py:
```diff
@@ -117,6 +124,13 @@
+    Args:
+        input_values (`torch.Tensor` of shape `({0})`):
+            Float values of input raw speech waveform..
+            Input values can be obtained from audio file loaded into an array using [`AutoFeatureExtractor`](https://huggingface.co/docs/transformers/autoclass_tutorial#autofeatureextractor).
```
Will the link work?

mht-sharma replied:
yes

michaelbenayoun reviewed (approved):
LGTM!
Thanks @mht-sharma

fxmarty commented on optimum/onnxruntime/modeling_ort.py:
```diff
@@ -1545,6 +1559,341 @@ def _prepare_onnx_outputs(self, onnx_outputs):
         return outputs
 
 
+AUDIO_CLASSIFICATION_EXAMPLE = r"""
```
Probably you already have - but can you make sure the code snippet works?

mht-sharma replied:
Done

fxmarty commented on optimum/onnxruntime/modeling_ort.py:
```diff
@@ -1545,6 +1559,341 @@ def _prepare_onnx_outputs(self, onnx_outputs):
+
+@add_start_docstrings(
+    """
+    Onnx Model for sequence classification head on top (a linear layer over the pooled output) for tasks like
```
```suggestion
    Onnx Model with a sequence classification head on top (a linear layer over the pooled output) for tasks like
```

fxmarty commented on optimum/onnxruntime/modeling_ort.py:
```diff
@@ -1545,6 +1559,341 @@ def _prepare_onnx_outputs(self, onnx_outputs):
+            return SequenceClassifierOutput(logits=logits)
+
+
+CTC_EXAMPLE = r"""
```
same comment

fxmarty commented on optimum/onnxruntime/modeling_ort.py:
```diff
@@ -1545,6 +1559,341 @@ def _prepare_onnx_outputs(self, onnx_outputs):
+            return CausalLMOutput(logits=logits)
+
+
+AUDIO_XVECTOR_EXAMPLE = r"""
```
same

fxmarty commented on optimum/onnxruntime/__init__.py:
```diff
@@ -29,7 +29,11 @@
     ],
     "modeling_ort": [
         "ORTModel",
+        "ORTModelForAudioClassification",
+        "ORTModelForAudioFrameClassification",
+        "ORTModelForAudioXVector",
         "ORTModelForCustomTasks",
+        "ORTModelForCTC",
```
Can you add them in the documentation .mdx as well?

mht-sharma replied:
Done

fxmarty commented on optimum/onnxruntime/modeling_ort.py:
```diff
@@ -1545,6 +1559,341 @@ def _prepare_onnx_outputs(self, onnx_outputs):
+
+@add_start_docstrings(
+    """
+    Onnx Model for with a `language modeling` head on top for Connectionist Temporal Classification (CTC).
```
```suggestion
    Onnx Model with a language modeling head on top for Connectionist Temporal Classification (CTC).
```

fxmarty commented on tests/onnxruntime/test_modeling.py:
```diff
@@ -2431,6 +2439,416 @@ def test_compare_to_io_binding(self, model_arch):
+        model_id = self.ARCH_MODEL_MAP[model_arch] if model_arch in self.ARCH_MODEL_MAP else MODEL_NAMES[model_arch]
+        onnx_model = ORTModelForAudioClassification.from_pretrained(self.onnx_model_dirs[model_arch])
+
+        self.assertIsInstance(onnx_model.model, onnxruntime.capi.onnxruntime_inference_collection.InferenceSession)
+        self.assertIsInstance(onnx_model.config, PretrainedConfig)
+
+        set_seed(SEED)
+        transformers_model = AutoModelForAudioClassification.from_pretrained(model_id)
+        processor = get_preprocessor(model_id)
+        input_values = processor(self._generate_random_audio_data(), return_tensors="pt")
+        onnx_outputs = onnx_model(**input_values)
+
+        self.assertTrue("logits" in onnx_outputs)
+        self.assertIsInstance(onnx_outputs.logits, torch.Tensor)
+
+        with torch.no_grad():
+            transformers_outputs = transformers_model(**input_values)
+
+        # compare tensor outputs
+        self.assertTrue(torch.allclose(onnx_outputs.logits, transformers_outputs.logits, atol=1e-4))
+
+        gc.collect()
```
We can leave it to an other PR (as you want), but following https://github.com/huggingface/optimum/pull/790 we will need to accept numpy inputs/outputs and change the tests accordingly.

mht-sharma replied:
will do in other PR

fxmarty replied:
Why not in this PR? It shouldn't be much work.

mht-sharma replied:
Done

mht-sharma replied:
dcs

fxmarty commented on tests/onnxruntime/test_modeling.py:
```diff
@@ -2431,6 +2439,416 @@ def test_compare_to_io_binding(self, model_arch):
+    )
+    @require_torch_gpu
+    @pytest.mark.gpu_test
+    def test_pipeline_on_gpu(self, test_name: str, model_arch: str, provider: str):
```
Could you make sure the tests pass on GPU?

fxmarty reviewed (approved):
Great PR, thank you! I left a few minor comments. As long as the tests (including GPU ones) pass, it is good to me.

fxmarty commented:
> The PR adds:
>     ORTModelForAudioClassification
>     ORTModelForCTC
>     ORTModelForAudioXVector
>     ORTModelForAudioFrameClassification

Is there a reference in transformers doc what these correspond to? For the layman (like me), it can be tricky to know. Maybe it can be good to point in the doc (or docstrings of the class) a good reference for each in the transformers doc, if any exists.

mht-sharma commented:
> > The PR adds:
> > ORTModelForAudioClassification
> > ORTModelForCTC
> > ORTModelForAudioXVector
> > ORTModelForAudioFrameClassification
> 
> Is there a reference in transformers doc what these correspond to? For the layman (like me), it can be tricky to know. Maybe it can be good to point in the doc (or docstrings of the class) a good reference for each in the transformers doc, if any exists.

Added reference to `AutoModel`

> > The PR adds:
> > ORTModelForAudioClassification
> > ORTModelForCTC
> > ORTModelForAudioXVector
> > ORTModelForAudioFrameClassification
> 
> Is there a reference in transformers doc what these correspond to? For the layman (like me), it can be tricky to know. Maybe it can be good to point in the doc (or docstrings of the class) a good reference for each in the transformers doc, if any exists.

Added reference to `AutoModelXXX`

fxmarty commented on docs/source/onnxruntime/package_reference/modeling_ort.mdx:
```diff
@@ -66,3 +66,19 @@ specific language governing permissions and limitations under the License.
 
 [[autodoc]] onnxruntime.ORTModelForVision2Seq
+
+## ORTModelForAudioClassification
```
For now it was ordered alphabetically, but maybe it makes sense to have categories by modality (text, audio, image)

mht-sharma replied:
Added

fxmarty commented on optimum/onnxruntime/modeling_ort.py:
```diff
@@ -1648,6 +1662,352 @@ def _prepare_onnx_inputs(self, use_torch: bool, **kwargs):
+
+@add_start_docstrings(
+    """
+    Onnx Model for with an XVector feature extraction head on top for tasks like Speaker Verification.
```
```suggestion
    Onnx Model with an XVector feature extraction head on top for tasks as speaker verification.
```

fxmarty commented on optimum/onnxruntime/modeling_seq2seq.py:
```diff
@@ -300,6 +307,12 @@ def forward(
+            if "attention_mask" in self.input_names:
+                onnx_inputs["attention_mask"] = attention_mask.cpu().detach().numpy()
+
+                # TODO: Replace with a better solution
+                if self.session.get_inputs()[1].type == "tensor(int64)":
+                    onnx_inputs["attention_mask"] = onnx_inputs["attention_mask"].astype(np.int64)
```
Why is this needed?

mht-sharma replied:
Some of the model tokenisers generate input in `int32` but the model export always happens with `int64` dtype (int input generator always generates this input).  This cause an error during inference because of incorrect datatype.

fxmarty replied:
Could you add a comment for this?

mht-sharma replied:
Added

fxmarty commented on optimum/onnxruntime/modeling_ort.py:
```diff
@@ -1648,6 +1662,352 @@ def _prepare_onnx_inputs(self, use_torch: bool, **kwargs):
+        **kwargs,
+    ):
+        if self.device.type == "cuda" and self.use_io_binding:
+            io_binding, output_shapes, output_buffers = self.prepare_io_binding(
+                input_values, ordered_input_names=self._ordered_input_names
+            )
+
+            # run inference with binding & synchronize in case of multiple CUDA streams
+            io_binding.synchronize_inputs()
+            self.model.run_with_iobinding(io_binding)
+            io_binding.synchronize_outputs()
+
+            # converts output to namedtuple for pipelines post-processing
+            return TokenClassifierOutput(logits=output_buffers["logits"].view(output_shapes["logits"]))
```
So if I read correctly, this is not tested in test_modeling.py. Could we rather raise a `NotImplementedError` here?

mht-sharma replied:
Sure
